diff --git a/src/xwayland/Server.cpp b/src/xwayland/Server.cpp
index e2dab412..309d4837 100644
--- a/src/xwayland/Server.cpp
+++ b/src/xwayland/Server.cpp
@@ -298,7 +298,7 @@ void CXWaylandServer::runXWayland(int notifyFD) {
         _exit(EXIT_FAILURE);
     }
 
-    auto cmd = std::format("Xwayland {} -rootless -core -listenfd {} -listenfd {} -displayfd {} -wm {}", displayName, xFDs[0], xFDs[1], notifyFD, xwmFDs[1]);
+    auto cmd = std::format("Xwayland {} -rootless -core -terminate 10 -listenfd {} -listenfd {} -displayfd {} -wm {}", displayName, xFDs[0], xFDs[1], notifyFD, xwmFDs[1]);
 
     auto waylandSocket = std::format("{}", waylandFDs[1]);
     setenv("WAYLAND_SOCKET", waylandSocket.c_str(), true);
diff --git a/src/xwayland/XWM.cpp b/src/xwayland/XWM.cpp
index b34d0cfe..f4df7682 100644
--- a/src/xwayland/XWM.cpp
+++ b/src/xwayland/XWM.cpp
@@ -829,6 +829,8 @@ CXWM::CXWM() {
     };
     xcb_change_property(connection, XCB_PROP_MODE_REPLACE, screen->root, HYPRATOMS["_NET_SUPPORTED"], XCB_ATOM_ATOM, 32, sizeof(supported) / sizeof(*supported), supported);
 
+    xcb_xfixes_set_client_disconnect_mode(connection, XCB_XFIXES_CLIENT_DISCONNECT_FLAGS_TERMINATE);
+
     xcb_flush(connection);
 
     setActiveWindow(XCB_WINDOW_NONE);
